### PR TITLE
Fixes for a couple of types of animation problems we encountered

### PR DESCRIPTION
**Pull request by elenadBlueSky**
Fixes for a couple of types of animation problems we encountered:
1. Import a usd file containing an animated camera, save to a Maya file,
   reopen Maya file, animation is gone.
2. Import a usd file with some geometry, keyframe animation on some of the
   elements, save and reopen Maya file. Animation is missing plus user
   can no longer interactively transform some of the elements with Maya's
   manipulators.
One of the issues appeared to be the fact that "over"s were being added in
usd in the edit layer which overrode the animation data in the earlier layers.
Another issue looked to be the fact that an animated element was purposefully
being considered read-only - this was not allowing the user to move it.

### CODE REVIEW

File: plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp [hunk 1]
```diff
@@ -872,37 +896,44 @@ void TransformationMatrix::initialiseToPrim(bool readFromPrim, Transform* transf
...
+
     switch(*opIt)
     {
     case kTranslate:
       {
         m_flags |= kPrimHasTranslation;
-        if(op.GetNumTimeSamples() > 1)
-        {
+        if(numTimeSamples > 0)
           m_flags |= kAnimatedTranslation;
```
Comment (elrond79) [inline plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp]: I see that you're changing the condition here, from > 1 to > 0.

This makes the behavior in `initialiseToPrim` match that in `updateToTime`, which is good... but I'm wondering what your rationale was for changing everything to > 0, as opposed to making everything be > 1.  At a guess, it seems like > 1 might make more sense as a criteria for whether there's animation, since with only 1 I would assume the value would still be static.

Might also be good to hear from someone at AL on whether there was a reason for the different criteria in updateToTime vs initialisePrim, or whether that was just an oversight.  @murphyeoin  or @robthebloke ?

File: plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp [hunk 2]
```diff
@@ -1263,7 +1309,12 @@ MStatus TransformationMatrix::translateTo(const MVector& vector, MSpace::Space s
     {
       insertTranslateOp();
     }
-    pushTranslateToPrim();
+
+    // Push new value to prim, but only if it's changing, otherwise extra work and unintended
+    // side effects will happen.
+    //
+    if (!apxEq(vector, m_translationFromUsd))
```
Comment (elrond79) [inline plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp]: First - as noted above, perhaps we should use vector::isEquivalent here.  I also think that it's default (1e-10) is a little safer, since a value of 1e-4 could easily be noticeable, especially if there's a scene that has smaller-than-average scale.  (Shouldn't happen, but sometimes does.)  If 1e-10 is still letting through changes essentially just due to floating point roundoff error to sneak through... maybe we could try 1e-8?

Also, you're not accounting for the tweak value here.  I guess simplest way to get the same behavior as current - while avoiding unnecessary over creation - would be something like:

```
if(apEx(..))
{
  m_translationTweak = MVector(0, 0, 0);
}
else
{
  pushTranslateToPrim();
}
```

**Comment by robthebloke** [comment]
Yeah, this PR re-introduces a nasty bug. Load this file into a proxy shape, select the cube, enable the translate tool, and then scrub the animation. The TRS manips no longer follow the prim. 
```
#usda 1.0
(
    defaultPrim = "pCube1"
    endTimeCode = 20
    startTimeCode = 1
)

def Mesh "pCube1"
{
    int[] faceVertexCounts = [4, 4, 4, 4, 4, 4]
    int[] faceVertexIndices = [0, 1, 3, 2, 2, 3, 5, 4, 4, 5, 7, 6, 6, 7, 1, 0, 1, 7, 5, 3, 6, 0, 2, 4]
    point3f[] points = [(-0.5, -0.5, 0.5), (0.5, -0.5, 0.5), (-0.5, 0.5, 0.5), (0.5, 0.5, 0.5), (-0.5, 0.5, -0.5), (0.5, 0.5, -0.5), (-0.5, -0.5, -0.5), (0.5, -0.5, -0.5)]
    float3 xformOp:rotateXYZ:rotate.timeSamples = {
        1: (0, 0, 0),
        2: (-0.20368686, -0.025981952, 0.65669316),
        3: (-0.78512025, -0.10014861, 2.5312536),
        4: (-1.6998594, -0.2168312, 5.4804025),
        5: (-2.9034636, -0.3703609, 9.360863),
        6: (-4.3514924, -0.55506897, 14.029353),
        7: (-5.9995036, -0.76528656, 19.342598),
        8: (-7.803058, -0.99534494, 25.157316),
        9: (-9.717715, -1.2395753, 31.330233),
        10: (-11.699033, -1.4923089, 37.718067),
        11: (-13.70257, -1.7478766, 44.177536),
        12: (-15.683888, -2.0006104, 50.565372),
        13: (-17.598545, -2.2448406, 56.738285),
        14: (-19.4021, -2.474899, 62.553005),
        15: (-21.05011, -2.6851168, 67.86625),
        16: (-22.49814, -2.8698246, 72.534744),
        17: (-23.701744, -3.0233545, 76.4152),
        18: (-24.616484, -3.1400368, 79.36436),
        19: (-25.197918, -3.2142036, 81.238914),
        20: (-25.401604, -3.2401857, 81.8956),
    }
    float3 xformOp:scale:scale = (1, 1, 1)
    float3 xformOp:translate:translate.timeSamples = {
        1: (0, 0, 0),
        2: (-0.020815758, 0.02076946, 0.0052386625),
        3: (-0.08023528, 0.080056824, 0.020192662),
        4: (-0.17371696, 0.17333058, 0.04371902),
        5: (-0.29671916, 0.29605922, 0.074674755),
        6: (-0.44470027, 0.4437112, 0.11191688),
        7: (-0.61311865, 0.611755, 0.15430242),
        8: (-0.7974327, 0.7956591, 0.20068839),
        9: (-0.9931009, 0.99089205, 0.24993183),
        10: (-1.1955814, 1.1929222, 0.30088973),
        11: (-1.4003328, 1.3972182, 0.3524191),
        12: (-1.6028134, 1.5992484, 0.403377),
        13: (-1.7984815, 1.7944813, 0.45262045),
        14: (-1.9827955, 1.9783854, 0.49900642),
        15: (-2.151214, 2.1464293, 0.54139197),
        16: (-2.299195, 2.2940812, 0.5786341),
        17: (-2.422197, 2.4168098, 0.6095898),
        18: (-2.515679, 2.5100837, 0.6331162),
        19: (-2.5750985, 2.569371, 0.64807016),
        20: (-2.5959141, 2.5901403, 0.6533088),
    }
    uniform token[] xformOpOrder = ["xformOp:translate:translate", "xformOp:rotateXYZ:rotate", "xformOp:scale:scale"]
}
```

File: plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp [hunk 3]
```diff
@@ -872,37 +896,44 @@ void TransformationMatrix::initialiseToPrim(bool readFromPrim, Transform* transf
     m_flags |= kInheritsTransform;
 
   if(matchesMayaProfile(m_xformops.begin(), m_xformops.end(), m_orderedOps.begin()))
-  {
     m_flags |= kFromMayaSchema;
```
[the remark this answers is not in the document]
Comment (elenadBlueSky) [reply plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp]: I'm completely fine with putting those back.

File: plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp [hunk 4]
```diff
@@ -872,37 +896,44 @@ void TransformationMatrix::initialiseToPrim(bool readFromPrim, Transform* transf
...
-  else
-  {
-  }
+
+  // Disable push to prim if enabled, otherwise MPlug value queries and setting
+  // in the switch statement below will trigger pushing to the prim, which
+  // will create undesirable usd "over"s.
+  //
+  bool pushOriginallyEnabled = pushToPrimEnabled();
+  m_flags &= ~kPushToPrimEnabled;
```
[the remark this answers is not in the document]
Comment (elenadBlueSky) [reply plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp]: Could you explain what a RAll object is?

File: plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp [hunk 5]
```diff
@@ -872,37 +896,44 @@ void TransformationMatrix::initialiseToPrim(bool readFromPrim, Transform* transf
...
   for(std::vector<UsdGeomXformOp>::const_iterator it = m_xformops.begin(), e = m_xformops.end(); it != e; ++it, ++opIt)
   {
     const UsdGeomXformOp& op = *it;
+
+    // How many time samples does the op have?  (This will be 0 if the op only has a constant
+    // value set or if it has no value set at all.)
+    //
+    int numTimeSamples = op.GetNumTimeSamples();
+
     switch(*opIt)
```
[the remark this answers is not in the document]
Comment (elenadBlueSky) [reply plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp]: This is just my personal style.  If some operation is repeated multiple times (even if those times are all in different conditionals) I tend to do the operation in one place and save/use the result.  (Also I had a debug print statement here for a while printing out the number of time samples and wanted to have only one of those.)  That said, I'm completely fine with putting the multiple calls to GetNumSamples back if you guys prefer that.

File: plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp (hunk 1 above)
Comment (elenadBlueSky) [reply plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp]: Since there is an explicit way in usd to set a constant value (e.g. "float3 xformOp:translate = (1,2,3)"), and a different way is used for time samples (e.g. "float3 xformOp:translate.timeSamples = {1: (1,2,3), 2: (..."), even if there is only one time sample it seems like the setting with time samples should be considered animation, as additional time samples may be added at any time.
In trying to address the problems we were encountering (unfortunately without having a good understanding of why the code was set up how it was and whether there were really good or maybe not so good reasons for it), I had to try a variety of things.  This was one of the tweaks that made intuitive sense to me and seemed to help so I kept it.

File: plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp [hunk 6]
```diff
@@ -872,37 +896,44 @@ void TransformationMatrix::initialiseToPrim(bool readFromPrim, Transform* transf
...
-            MPlug(transformNode->thisMObject(), MPxTransform::translateX).setValue(tempTranslation.x);
-            MPlug(transformNode->thisMObject(), MPxTransform::translateY).setValue(tempTranslation.y);
-            MPlug(transformNode->thisMObject(), MPxTransform::translateZ).setValue(tempTranslation.z);
+            MPlug(transformNode->thisMObject(), MPxTransform::translateX).setValue(vecFromUsd.x);
+            MPlug(transformNode->thisMObject(), MPxTransform::translateY).setValue(vecFromUsd.y);
+            MPlug(transformNode->thisMObject(), MPxTransform::translateZ).setValue(vecFromUsd.z);
             m_translationTweak[0] = m_translationTweak[1] = m_translationTweak[2] = 0;
-            m_translationFromUsd = tempTranslation;
+            m_translationFromUsd = vecFromUsd;
           }
```
[the remark this answers is not in the document]
Comment (elenadBlueSky) [reply plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp]: I added temp variables in all the cases of the switch and did this particular rename for consistency with those.  I named them all based on the data types.  (I had originally thought about declaring all the temp variables outside the switch and reusing them to make the switch cases have a little less stuff in them, but then didn't do that to avoid having some unused variables sitting around.)

File: plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp [hunk 7]
```diff
@@ -913,16 +944,17 @@ void TransformationMatrix::initialiseToPrim(bool readFromPrim, Transform* transf
         m_flags |= kPrimHasPivot;
         if(readFromPrim)
         {
-          internal_readPoint(m_scalePivotFromUsd, op);
-          m_rotatePivotFromUsd = m_scalePivotFromUsd;
+          MPoint ptFromUsd;
+          internal_readPoint(ptFromUsd, op);
           if(transformNode)
```
[the remark this answers is not in the document]
Comment (elenadBlueSky) [reply plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp]: I'm happy to add tweak zeroing to all the cases - this is one of those things that seems important that I didn't know because I don't have a deep understanding of the code I'm changing.
The reason I started using the temp variables everywhere is I noticed, as you said, callbacks being triggered at inopportune times with the result of undesirable USD "over" settings of values being added that covered up existing animation with over-ed constant values.  A number of the code changes in this pull request are to try and prevent callbacks from triggering too many times or at times that cause problems.
(Note that I also noticed an obviously wrong tweak value being auto-computed and set a few times while I was trying to figure out the code, so there may be some problem hiding in the tweak mechanism.  Unfortunately I don't have a reproducible case of this for you.)

File: plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp (hunk 2 above)
Comment (elenadBlueSky) [reply plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp]: I'm happy to swap my non-standard approximate equality checking for some more standard function - I didn't know what was available and figured that if there was something better someone would comment and point me at the better way.
Regarding your tweak comment, could you please elaborate on what you mean?  Are you saying that I need to set tweak to zero if not updating m_translationFromUsd?  Wouldn't it already be correctly set in this case?

File: plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp [hunk 8]
```diff
@@ -1118,12 +1157,15 @@ void TransformationMatrix::initialiseToPrim(bool readFromPrim, Transform* transf
...
+  // (Note - I don't think we want to trigger an update of anything here, only
+  // reenable push for future operations, hence not calling enablePushToPrim.)
+  //
+  if(pushOriginallyEnabled)
+    m_flags |= kPushToPrimEnabled;
+
   if(m_flags & kAnimationMask)
-  {
-    m_flags &= ~kPushToPrimEnabled;
     m_flags |= kReadAnimatedValues;
```
[the remark this answers is not in the document]
Comment (elenadBlueSky) [reply plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp]: In our problem case the user had animated the translations/rotations on some models in her scene and saved her Maya file.  When she reopened the Maya file later to continue working, she could select those models but could no longer translate them - she would move the manipulator, but that's all that would move, the model stayed where it was.  Commenting out the pushToPrim disabling took care of this.

**Comment by elenadBlueSky** [comment]
@robthebloke - I saved the usda file above, then imported it into Maya with maya.cmds.AL_usdmaya_ProxyShapeImport.  I selected the cube in the viewport, which caused temporary geometry to be created.  I selected, at various times, the Move Tool, Rotate Tool, and Scale Tool and clicked through the timeline.  All the manipulators move along with the moving cube.  The Rotate and Scale manipulators both translate and rotate, the Move manipulator just translates, but I thought it was normal behavior for it to align with the coordinate axes.  The manipulators disappear while I'm scrubbing through the timeline but reappear when I let go of the mouse on a frame.
Just to check, I created a nurbs cube object and set keys on two frames to animate its translation and rotation.  The way the manipulators behave on this Maya geometry is exactly the same as on the usd cube when I click along and scrub along the timeline.  Maybe I'm missing something?

**Comment by murphyeoin** [comment]
Hi @elenadBlueSky - @robthebloke will be on holiday until early in the new year so probably won't respond til then. I can see your point about the behaviour of the manips in USD/pure maya being the same - I'll get Rob to post an animated Gif to explain a bit better, along with some repro details (version of Maya, any special Viewport settings etc), or at least clarify what the difference in the two is

File: plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (murphyeoin) [reply plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp]: In this case It would mean creating an object that does the cleanup when it goes out of scope - just by having the same cleanup code as below in the destructor. This would be triggered when the code block complete (i.e normal execution) but also cover the case when an exception is thrown

[the remark this answers is not in the document]
Comment (elrond79) [reply plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp]: Yup, what Eoin said - more details here: https://en.cppreference.com/w/cpp/language/raii
The acronym is pretty opaque, but basically because c++ doesn’t have a try / finally statement, it’s a trick to use the deconstructor to serve the same purpose, some those are supposed to run even when unwinding the stack for exceptions.

[the remark this answers is not in the document]
Comment (elrond79) [reply plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp]: Yup, what Eoin said - more details here: https://en.cppreference.com/w/cpp/language/raii
The acronym is pretty opaque, but basically because c++ doesn’t have a try / finally statement, it’s a trick to use the deconstructor to serve the same purpose, some those are supposed to run even when unwinding the stack for exceptions.

File: plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (elrond79) [reply plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp]: Yeah, if it’s just a matter of personal style, I’d prefer to introduce fewer changes (and the old way is technically more efficient, though I’d guess GetNumTimeSamples is fast, so probably practically speaking that doesn’t matter.)

File: plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp (hunk 1 above)
Comment (elrond79) [reply plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp]: Hmm, well that argument makes sense to me. We’ll stick with your > 0 unless else objects...

File: plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (elrond79) [reply plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp]: Hmm, do you know how / why using the temp variables prevents overs being created?

**Comment by pilarmolinalopez** [comment]
hi @murphyeoin and @robthebloke - did you guys have some time to look into this? Thanks!

**Comment by kxl-adsk** [comment]
@pilarmolinalopez, @elenadBlueSky we have some unresolved comments in here requesting changes. Are you waiting for AL before addressing them?

**Comment by murphyeoin** [comment]
HI @pilarmolinalopez will look at this again with @robthebloke first thing monday

**Comment by pilarmolinalopez** [comment]
thanks, @murphyeoin !

hi @kxl-adsk - yes, I believe Elena is waiting for AL in case we have more changes to make. That being said we can also address those first comments right away if you prefer that.

**Comment by elrond79** [comment]
One thing that can be done regardless of what AL says, and I think is really needed, is adding a test for the creation of the over.  This will be good both to prevent future regressions, and help evaluate this PR.  ie, I think the "over" creation should be avoided by the [disabling / re-enabling of pushToPrim](https://github.com/Autodesk/maya-usd/pull/183/commits/d4251fe891dad4e8c5e2ee0aea3e79406167f4a1#r360073243), but the [justification for creating temp vars](https://github.com/Autodesk/maya-usd/pull/183/commits/d4251fe891dad4e8c5e2ee0aea3e79406167f4a1#r360493999) was that it also prevented these "overs" - I'd like to see if that is actually the case.

**Comment by elrond79** [comment]
Also, in general, having tests will help us pare down this PR to the bare minimum - this is important because both [AL](https://github.com/Autodesk/maya-usd/pull/158) and [Luma](https://github.com/Autodesk/maya-usd/compare/dev...LumaPictures:pr/al/use_UsdMayaXformStack) have some extensive changes to transformation code as well, so it would be nice to cut down on unnecessary merge conflicts.

**Comment by pilarmolinalopez** [comment]
Of course. I completely agree.

**Comment by kxl-adsk** [comment]
Thanks for the updates. My goal was to get a better understanding of where we are with this PR. Having said that, I would echo the idea of creating unit test/s.

**Comment by murphyeoin** [comment]
Hello @pilarmolinalopez @elenadBlueSky sorry we haven't unblocked this one yet -  I discussed with @robthebloke and there was the suggestion of putting the new behaviour behind an optionVar to begin with? I think though it might be good to have a chat about this on VC to try and resolve it more quickly... potentially with a maya session running. What do you think?

**Comment by pilarmolinalopez** [comment]
hi @murphyeoin , I agree having a chat about this is a good idea. I'll talk to @cfmoore007 to see how we can set it up. Thanks!

File: plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp [hunk 9]
```diff
@@ -34,6 +34,30 @@ namespace nodes {
...
+static bool apxEq(const MPoint &val1, const MPoint &val2, double epsilon=0.00001)
+{
+  return apxEq(val1.x, val2.x, epsilon) && apxEq(val1.y, val2.y, epsilon) && apxEq(val1.z, val2.z, epsilon);
+}
+
+static bool apxEq(const MEulerRotation &val1, const MEulerRotation &val2, double epsilon=0.00001)
+{
+  return apxEq(val1.x, val2.x, epsilon) && apxEq(val1.y, val2.y, epsilon) && apxEq(val1.z, val2.z, epsilon);
+}
+
```
[the remark this answers is not in the document]
Comment (elenadBlueSky) [reply plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp]: Replaced all approximate equality checks with calls to the isEquivalent method.  (I haven't submitted the changes yet, plan to do so later today.)

File: plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (elenadBlueSky) [reply plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp]: Put all the braces back.  Added braces in a few more places where they were missing, too. :)

File: plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (elenadBlueSky) [reply plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp]: Instead of complicating things with an extra object, I just used a try/catch block where the catch clause resets the original state and rethrows the exception in case something else is looking for it.  Hopefully that's acceptable.

File: plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (elenadBlueSky) [reply plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp]: I got rid of the numTimeSamples variable and put all the calls to GetNumTimeSamples() back.

File: plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (elenadBlueSky) [reply plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp]: Added tweak zeroing to all the cases.
Regarding why using the temp variables prevents overs being created, my guess was that when the internal_read* methods were reading into the non-temp variables, they were triggering some callback because the object was actually being changed.  For some reason, using MPlug.setValue to update the object later wasn't having the same effect.  Not sure why, but I was printing out messages in callbacks to watch what was being triggered when and noticed behavior differences.  It's been a while so I don't remember all the details.

File: plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp (hunk 2 above)
Comment (elenadBlueSky) [reply plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp]: I didn't do anything with the tweak because no one explained why resetting it was necessary here.  Will be happy to revisit.

**Comment by elenadBlueSky** [comment]
FYI - we tested our problem cases against the AL and Luma pull requests last week and those do not address them, so we feel our changes are still needed.

File: plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (pilarmolinalopez) [reply plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp]: When I first introduced the temp variables (back in the old AL repo days) the issue was that TransformationMatrix::pushToPrim() was looping over all the transform operators, pushing those values to the primitive and updating the m_* variables (and they would lose their values). For instance, when the translateX plug was being set, TransformationMatrix::pushToPrim() would be called, then MPxTransformationMatrix::translationValue would be pushed to the primitive and m_translationFromUsd updated with that value. However, at that moment MPxTransformationMatrix::translationValue only has the value read for the X component of the translation - Y and Z have not been set yet. The values read for Y and Z got lost

File: plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (elenadBlueSky) [reply plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp]: Changes submitted now.

File: plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (elrond79) [reply plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp]: Ok - so that definitely explains why this would be necessary if pushToPrim was still enabled.

However - I think that now that pushToPrim is disabled during this function, I think this shouldn't be necessary - and the simpler "write directly to m_scalePivotFromUsd" method should work.

Of course, knowing things like this "for sure" is precisely why I'd like to see tests for this!

[the remark this answers is not in the document]
Comment (elenadBlueSky) [reply plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp]: @pilarmolinalopez - Now that you mention it, I'd also run intothe same strange problem!  Only one of the xyz components of a vector was correctly updated at some point, so when it came time for the code to check if the saved vector was the same as the value we were trying to set it to, the answer came back that no it wasn't the same.  This resulted in some operation being performed that wouldn't have been performed if the x y and z had all been properly updated at the same time.

File: plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (elrond79) [reply plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp]: Ok - I can see how disabling pushToPrim might lead to the behavior you're describing, which definitely isn't desired.
However, I'm a little worried that NOT disabling it might also make the tweak not work as I assume it was intended to allow the maya user to apply a constant offset on top of the animation values coming from USD.  Personally, though, I've never made use of this - or even confirmed that this is, indeed, what it was supposed to do... and that it was working.
Perhaps we can get someone from AL to comment here?

**Comment by elenadBlueSky** [comment]
Regarding a unit test for "over" creation, I'm happy to add that but I'm going to need some guidance from you guys as I'm not clear on what exactly we would want the test to check for.  Maybe we can discuss this whenever the phone call gets scheduled.

File: plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp [hunk 10]
```diff
@@ -1261,7 +1317,14 @@ MStatus TransformationMatrix::translateTo(const MVector& vector, MSpace::Space s
...
     }
-    pushTranslateToPrim();
+
+    // Push new value to prim, but only if it's changing, otherwise extra work and unintended
+    // side effects will happen.
+    //
+    if (!vector.isEquivalent(m_translationFromUsd))
+    {
+      pushTranslateToPrim();
+    }
```
Comment (elrond79) [inline plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp]: So, this is a follow-up to [this thread](https://github.com/Autodesk/maya-usd/pull/183#discussion_r377780255), because I can't find a way to add more replies to an "outdated" review...

Anyway - I was mostly advocating zeroing the tweak because that's what the existing code would do, if pushTranslateToPrim were called. However - you're totally right, I now realize that this extra step is unnecessary, since [this line](https://github.com/Autodesk/maya-usd/blob/37bf63327f18c7944e30f715d9f061b4338e5c75/plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp#L1305)  is already zeroing tweak, if vector == m_translationFromUsd!

File: plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (elrond79) [reply plugin/al/lib/AL_USDMaya/AL/usdmaya/nodes/TransformationMatrix.cpp]: I'm going to mark this resolved since, if we do need / want temp vars for all these branches, then standardizing the naming does make sense.  However, as I note [here](https://github.com/Autodesk/maya-usd/pull/183#discussion_r377866663), I'm still not sure that the temp vars are really necessary...

**Comment by elrond79** [comment]
> Regarding a unit test for "over" creation, I'm happy to add that but I'm going to need some guidance from you guys as I'm not clear on what exactly we would want the test to check for. Maybe we can discuss this whenever the phone call gets scheduled.

Well, I guess for over creation, I guess just a test that would set the translation to the same value (or maybe one VERY close, within the episilon tolerance), and just confirm that no over was added to the usd.  (Not sure offhand the best way to check that you - might have to check the primspec at the layer level?  Or, assuming it was set to something nearly-the-same, we could just check that the compsed value didn't change.)

I'd also like to see tests for the two scenarios laid out at the top, in the description of the PR.

**Comment by elrond79** [comment]
Also - I'm curious, it sounds like what you folks are trying to do is create animation, using usd-transforms as the "controls".

Is that indeed what you're attempting? Does that workflow work with your changes? Is the idea that you'd maya maya animation curves, and then export those to USD?  Or is the end goal to somehow animate "straight" to USD, by setting time-samples directly (something the plugin doesn't currently support)?

Would love to hear your results - we've only ever attempted to do animation via the MayaReference schema, to bring in "real" maya rigs in the appropriate place in the USD hierarchy. But I can see how animating directly on the USD transforms could be useful in some situations!

**Comment by pilarmolinalopez** [comment]
Hi Paul - we are indeed trying to have animated AL transforms for our basic layout workflow :)

We also rely on the MayaReference schema to animate via maya rigs that we bring in by switching variants in the proxy shape. We'll be happy to chat about all this if you want more details.

**Comment by elrond79** [comment]
Yeah, that might be good - we're currently in the middle of deciding where
we want to go with our current USD layout workflow...

**Comment by kxl-adsk** [comment]
@elenadBlueSky With #294 and #295 now open, can we close this PR?

**Comment by elenadBlueSky** [comment]
#294 and #295 only contain part of what's in #183 - they were the less controversial bits.  The AL developers were going to look at the more controversial parts of #183 and try to come up with alternative solutions to the problem I was trying to address with those parts.  So maybe wait a bit before closing 183 unless the closed PR will still give them a way to reference the changes I made and everyone's comments above?

**Comment by kxl-adsk** [comment]
You can still view and comment on closed PRs - example https://github.com/Autodesk/maya-usd/pull/252

**Comment by pilarmolinalopez** [comment]
hi @kxl-adsk , I was talking to Elena about this. It's okay to close this PR. It makes bookkeeping easier. As you said, AL will still have access to it if they want to check out our changes. We may even submit a separate PR with the changes made here that are not covered by #294 and #295. Thanks!
